### DISCUSSION TITLE
PyODBC for PostgreSQL support

### DISCUSSION THREAD

**Discussion by tzing** [sqlalchemy/sqlalchemy#9959] [Usage Questions]
### Describe the use case

For short - Literally want to use PyODBC on PostgreSQL dialect.

I am using AWS RedShift, which speaks PSQL protocol. 
And my code runs in a protected environment, I cannot touch public network without proxy.

The feasible solution for me is to use AWS's official [ODBC driver](https://docs.aws.amazon.com/redshift/latest/mgmt/configure-odbc-connection.html), but currently we do not have PyODBC support on PSQL.

I know there is [sqlalchemy-redshift](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/) project - they bring [redshift-connector](https://pypi.org/project/redshift-connector/) to sqlalchemy. I didn't see an official document using proxy so it isn't work for me.

### Databases / Backends / Drivers targeted

- Database: AWS RedShift
- Dialect: PostgreSQL
- Driver: PyODBC

### Example Use

For general use cases:

```py
postgresql_engine = create_engine(
    "postgresql+pyodbc://scott:tiger@hostname/dbname",
)
```

For personal use case:

```py
import urllib
connection_string = (
    'DRIVER=/opt/amazon/redshift/lib/amazonredshiftodbc.dylib;'
    'SERVER=example.redshift.amazonaws.com;'
    'PORT=5439;'
    'DATABASE=mydb;'
    'UID=root;'
    'PWD=(whatever);'
    'ProxyHost=proxy;'
    'ProxyPort=3128;'
)
params = urllib.parse.quote_plus(connection_string)
connection_uri = "postgresql+pyodbc:///?odbc_connect=%s" % params
```

`ProxyHost` and `ProxyPort` are special connection string consumed by [Amazon Redshift ODBC Data Connector](https://s3.amazonaws.com/redshift-downloads/drivers/odbc/1.4.65.1000/Amazon+Redshift+ODBC+Connector+Install+Guide.pdf)

### Additional context

I feel it could be achieved by adopting [PyODBC connector](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/connectors/pyodbc.py) to PSQL dialect.

If no further concern, I could work on PR.

**Comment by zzzeek** [comment]
hi -

the pyodbc driver is very problematic for databases that aren't MSSQL and there should be no reason anyone needs this driver for PostgreSQL or AWS.      I am strong -1 on adding support for this driver.

What is the technical reason you can't use a proxy with psycopg2 ?

**Reply by zzzeek** [reply]
other options:

1. post an issue at sqlalchemy-redshift for proxy support and/or pyodbc support
2. post an issue at https://github.com/aws/amazon-redshift-python-driver requesting proxy support

these are all more direct places to get the feature adds you need

**Comment by tzing** [comment]
> What is the technical reason you can't use a proxy with psycopg2 ?

psycopg2 does not have proxy

> other options: ...

got it

**Reply by zzzeek** [reply]
a "proxy" is just a listening address somewhere, so depending on what protocol the "proxy" uses, you could connect any database adapter to it directly.   I dont know what kind of "proxy" this is.

the problem is this is some relatively short-term connectivity issue, adding pyodbc+pg directly to SQLAlchemy means I have to develop and maintain this code that virtually nobody will ever use in perpetuity.    pyodbc has a lot of problems with datatypes and the PG drivers do as well; putting them together would be a support nightmare.

you can of course make your own ad-hoc pyodbc+pg dialect just to connect and run simple queries without this being part of SQLAlchemy.   Just subclass [PyOBDCConnector](https://github.com/sqlalchemy/sqlalchemy/blob/98401bdcf46c2d200be82e275961fc5678c9950c/lib/sqlalchemy/connectors/pyodbc.py#L35) and [PGDialect](https://github.com/sqlalchemy/sqlalchemy/blob/98401bdcf46c2d200be82e275961fc5678c9950c/lib/sqlalchemy/dialects/postgresql/base.py#L2916) together.   See https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/dialects/mysql/pyodbc.py for how this is done for MySQL for example, it's very short
